Source/WebKit/WebProcess/Network/webrtc/LibWebRTCProvider.h

```diff
@@ -78,9 +78,13 @@ class LibWebRTCProvider final : public LibWebRTCProviderBase {
 #endif
 };
 
-inline UniqueRef<LibWebRTCProvider> createLibWebRTCProvider(WebPage& page)
+inline UniqueRef<LibWebRTCProviderBase> createLibWebRTCProvider(WebPage& page)
 {
-    return makeUniqueRef<LibWebRTCProvider>(page);
+    static bool useWebKitProvider = !!getenv("WPE_USE_WEBKIT_WEBRTC_PROVIDER");
+    if (useWebKitProvider) {
+       return makeUniqueRef<LibWebRTCProvider>(page);
```

Comment:
iiuc if this provider is used then WebRTC sockets will be created in the WebProcess instead of NetworkProcess and that would be a WebProcess sandbox violation.

Maybe the bottleneck you're seeing is related with the WebRTC IPC between the NetworkProcess and WebProcess?

Reply:
@philn this change was taken from one of our custom patches for older webkit 2.22.
Back then is it was concluded that CPU was wasted for NetworkProcess<->WebProcess IPC and samples passing through main thread, something like this:
```
     auto sample = GStreamerSampleFromLibWebRTCVideoFrame(frame);
+    #if 1
+    videoSampleAvailable(MediaSampleGStreamer::create(sample.get(), WebCore::FloatSize(), String()));
+    #else
     callOnMainThread([protectedThis = makeRef(*this), sample] {
         protectedThis->videoSampleAvailable(MediaSampleGStreamer::create(sample.get(),
             WebCore::FloatSize(), String()));
     });
```
The later seems to be not a case any more with 2.38 but IPC is still there.
I still need to look into this as working around IPC seems to have other side effects in video quality so there must be something more.

Anyway, any suggestions how to improve CPU usage will be welcome.

Reply:
Is that code from Source/WebCore/platform/mediastream/libwebrtc/gstreamer/RealtimeIncomingVideoSourceLibWebRTC.cpp ? There's no callOnMainThread there upstream.

Reply:
Right, This was for webkit 2.22. In 2.38 it was reworked with https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/da51c9d94b57995c7d92f8086a60a731b562f2da

Reply:
I know nothing about webprocess sandbox to be honest. We have dobby containers at upper level so I doubt this is enabled. The change works fine for me at least.
I can verify if you need

Reply:
Let's add a comment here saying the WebProcess sandbox is disabled, so moving LibWebRTC networking out of it doesn't make much sense... And remove the getenv call. This won't be suitable for upstream.

Reply:
ok, new patch uplaoded, please take a look if that is what you requested.

I also tried your new changes and everything works like wonder now. CPU usage of whole app is 70-80%, that is almost a half of what we had initially. No package loss observed.